pytests/test_types.py

```diff
@@ -0,0 +1,458 @@
+        code="ABC",
+        code_type=delivery_area_pb2.EnergyMarketCodeType.ENERGY_MARKET_CODE_TYPE_EUROPE_EIC,
+    ),
+    delivery_period=delivery_duration_pb2.DeliveryPeriod(
+        start=START_TIME_PB,
+        duration=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+    ),
+    execution_time=EXECUTION_TIME_PB,
+    price=price_pb2.Price(
+        amount=decimal_pb2.Decimal(value="100.00"),
+        currency=price_pb2.Price.Currency.CURRENCY_USD,
+    ),
+    quantity=energy_pb2.Energy(mwh=decimal_pb2.Decimal(value="5.00")),
+    state=electricity_trading_pb2.TradeState.TRADE_STATE_ACTIVE,
+)
+
+GRIDPOOL_ORDER_FILTER = GridpoolOrderFilter(
+    order_states=[OrderState.ACTIVE, OrderState.CANCELED],
+    side=MarketSide.BUY,
+    delivery_period=DeliveryPeriod(start=START_TIME, duration=timedelta(minutes=15)),
+    delivery_area=DeliveryArea(code="XYZ", code_type=EnergyMarketCodeType.EUROPE_EIC),
+    tag="test",
+)
+GRIDPOOL_ORDER_FILTER_PB = electricity_trading_pb2.GridpoolOrderFilter(
+    states=[
+        electricity_trading_pb2.OrderState.ORDER_STATE_ACTIVE,
+        electricity_trading_pb2.OrderState.ORDER_STATE_CANCELED,
+    ],
+    side=electricity_trading_pb2.MarketSide.MARKET_SIDE_BUY,
+    delivery_period=delivery_duration_pb2.DeliveryPeriod(
+        start=START_TIME_PB,
+        duration=delivery_duration_pb2.DeliveryDuration.DELIVERY_DURATION_15,
+    ),
+    delivery_area=delivery_area_pb2.DeliveryArea(
+        code="XYZ",
+        code_type=delivery_area_pb2.EnergyMarketCodeType.ENERGY_MARKET_CODE_TYPE_EUROPE_EIC,
+    ),
+    tag="test",
+)
+
+UPDATE_ORDER = UpdateOrder(price=Price(amount=Decimal("100.00"), currency=Currency.USD))
+UPDATE_ORDER_PB = electricity_trading_pb2.UpdateGridpoolOrderRequest.UpdateOrder(
+    price=price_pb2.Price(
+        amount=decimal_pb2.Decimal(value="100.00"),
+        currency=price_pb2.Price.Currency.CURRENCY_USD,
+    )
+)
+
+
+def test_currency_from_pb() -> None:
+    """Test the currency conversion from protobuf to enum."""
+    assert Currency.from_pb(price_pb2.Price.Currency.CURRENCY_EUR) == Currency.EUR
+    assert (
+        Currency.from_pb(price_pb2.Price.Currency.CURRENCY_UNSPECIFIED)
+        == Currency.UNSPECIFIED
+    )
+
+
+def test_currency_to_pb() -> None:
+    """Test the currency conversion from enum to protobuf."""
+    assert Currency.EUR.to_pb() == price_pb2.Price.Currency.CURRENCY_EUR
+    assert Currency.UNSPECIFIED.to_pb() == price_pb2.Price.Currency.CURRENCY_UNSPECIFIED
+
+
+def test_price_to_pb() -> None:
+    """Test the client price type conversions to protobuf."""
+    original_price = Price(amount=Decimal("100"), currency=Currency.USD)
+    converted_price = original_price.to_pb()
+
+    assert isinstance(converted_price, price_pb2.Price)
+    assert converted_price.amount == decimal_pb2.Decimal(value="100")
+    assert converted_price.currency == price_pb2.Price.Currency.CURRENCY_USD
+
```

Comment:
This might be a longer and maybe unnecessary change but it is something to keep in mind for the future. There are a lot of conversions to pb in this script which could be abstracted in the future into helper functions to reduce duplications and make the code a bit more readable. So you could have a function called `assert_conversion_to_pb` which takes in the original type and has an input which is the function which converts the original to the protobuf representation and asserts it (the assert function can also be a callable input of the `assert_conversion_to_pb` function.

Like I said, this is maybe not a priority for now if we want to move fast now so I will leave it up to you if you want to implement it. But maybe something to keep in mind for the future either way.

Reply:
A little example of what I mean with the help of chatgpt:

```python
def assert_conversion_to_pb(original: Any, convert_to_pb: Callable[[], T], expected_pb: T, assert_func: Callable[[T, T], None]) -> None:
    """
    Generic utility function to test conversion from a custom type to protobuf.

    :param original: The original instance of the custom type.
    :param convert_to_pb: Function that converts `original` to its protobuf representation.
    :param expected_pb: The expected protobuf instance.
    :param assert_func: Function to assert equality between two protobuf instances.
    """
    converted_pb = convert_to_pb()
    assert_func(converted_pb, expected_pb)
```

Then we would need a specific assert function for comparing pb messages:

```python
def assert_protobuf_equal(actual_pb: Any, expected_pb: Any) -> None:
    """
    Asserts that two protobuf instances are equal, using DeepDiff for detailed comparison.

    :param actual_pb: The actual protobuf instance.
    :param expected_pb: The expected protobuf instance.
    """
    diff = DeepDiff(actual_pb, expected_pb, ignore_order=True)
    assert not diff, f"Differences found in protobuf comparison: {diff}"
```

and then we could refactor the test cases like this:

```python
def test_price_to_pb_refactored() -> None:
    """Refactored test for the client price type conversions to protobuf using DRY principle."""
    assert_conversion_to_pb(
        original=Price(amount=Decimal("100"), currency=Currency.USD),
        convert_to_pb=lambda: Price(amount=Decimal("100"), currency=Currency.USD).to_pb(),
        expected_pb=price_pb2.Price(
            amount=decimal_pb2.Decimal(value="100"),
            currency=price_pb2.Price.Currency.CURRENCY_USD,
        ),
        assert_func=assert_protobuf_equal
    )
```